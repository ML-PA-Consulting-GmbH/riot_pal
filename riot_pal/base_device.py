# Copyright (c) 2018 Kevin Weiss, for HAW Hamburg  <kevin.weiss@haw-hamburg.de>
#
# This file is subject to the terms and conditions of the MIT License. See the
# file LICENSE in the top level directory for more details.
# SPDX-License-Identifier:    MIT
"""Abstract Base Device for RIOT PAL
This module is the abstract device that interfaces to a driver.  It handles
the selection and initialzation of the driver.

Example:
    class ApplicationDevice(BaseDevice):
        def show_example(self):
            self._write('Send Example')
            print(self._read())

    example_use_case = ApplicationDevice('serial', port='my/port/name')
    example_use_case.show_example()
"""
import logging
try:
    from . import driver_manager
except SystemError:
    import driver_manager


class BaseDevice:
    """Instance for devices to connect and utilize drivers.

    Args:
        *args: Variable length argument list.
        **kwargs: Arbitrary keyword arguments.
    """

    def __init__(self, *args, **kwargs):
        self._driver = driver_manager.driver_from_config(*args, **kwargs)

    def close(self):
        """Closes the device connection."""
        self._driver.close()

    def _read(self):
        """Reads data from the driver.

        Returns:
            str: string of data if success, ERR string if failed
        """
        return self._driver.read()

    def _write(self, data):
        """Writes data to the driver.

        Args:
            data(str): Variable length argument list.
        """
        return self._driver.write(data)

    def is_connected_to_board(self):
        """Dummy - confirm if a connection is for target board."""
        logging.warning("Check if board is connected: dummy should be"
                        " implmeneted in subclasses")
        raise NotImplementedError()

    @classmethod
    def from_autodetect(cls, *args, **dev_config):
        """Connects to a range of possible configurations."""
        configs = driver_manager.available_configs(*args, **dev_config)
        logging.debug("Configs: %r", configs)
        for config in configs:
            for retry in range(0, 2):
                logging.debug("Autodetect attempt: %d", retry)
                conn = cls(**config)
                try:
                    if conn.is_connected_to_board():
                        return conn
                except Exception as err:
                    logging.debug("Cannot connect: %r", err)
                conn.close()

        raise ValueError("Could not locate board, check if board is"
                         "connected or is_connected_to_board is correct")

    @classmethod
    def copy_driver(cls, device):
        """Copies the driver instance so many devices can use one driver."""
        logging.debug("Cloning Driver: %r", device._driver)
        return cls(dev_type='driver', driver=device._driver)


def main():
    """Tests basic usage of the class

    Used for unit testing, information should be confirm with DEBUG info.
    """
    logging.getLogger().setLevel(logging.DEBUG)
    BaseDevice()


if __name__ == "__main__":
    main()
