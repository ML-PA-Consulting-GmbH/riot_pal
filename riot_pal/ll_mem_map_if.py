# Copyright (c) 2018 Kevin Weiss, for HAW Hamburg  <kevin.weiss@haw-hamburg.de>
#
# This file is subject to the terms and conditions of the MIT License. See the
# file LICENSE in the top level directory for more details.
# SPDX-License-Identifier:    MIT
"""Low Level Memory Map Interface for RIOT PAL
This module handles offset and sizes dictated a memory map.
"""
import logging
import csv
try:
    from .ll_shell import LLShell
except SystemError:
    from ll_shell import LLShell


class LLMemMapIf(LLShell):
    """Interface between ll_shell and mem_map."""

    def __init__(self, path, *args, **kwargs):
        self.cmd_list = LLMemMapIf.import_mm_from_csv(path)
        super().__init__(*args, **kwargs)

    @staticmethod
    def import_mm_from_csv(path):
        """Imports a memory map csv file."""
        cmd_list = {}
        with open(path) as csvfile:
            rows = list(csv.reader(csvfile, quotechar="'"))
            for row in range(1, len(rows)):
                cmd = dict(zip(rows[0], rows[row]))
                cmd_list[rows[row][rows[0].index('name')]] = cmd
                logging.debug("Imported command: %r", cmd)
        return cmd_list

    def read_struct(self, cmd_name):
        """Reads a set of registers defined by the memory map."""
        response = []
        for cmd, val in self.cmd_list.items():
            # Supress unused variable warning
            val = val
            if cmd.startswith(cmd_name):
                response.append(self.read_reg(cmd))
        return response

    def read_reg(self, cmd_name, offset=0, size=None):
        """Read a register defined by the memory map.

        Keyword arguments:
        offset -- the number of elements to offset in an array
        size   -- the number of elements to read in an array
        """
        cmd = self.cmd_list[cmd_name]
        response = None
        if size is None:
            size = cmd['total_size']
        if cmd['is_bitfield'] is 'True':
            response = self.read_bits(cmd['offset'],
                                      cmd['bit_offset'],
                                      cmd['bits'])
        elif cmd['size'] != cmd['total_size']:
            offset = int(cmd['offset']) + (offset * int(cmd['size']))
            if size is None:
                size = cmd['total_size']
            else:
                size = size * int(cmd['size'])
            response = self.read_bytes(offset, size)
        else:
            response = self.read_bytes(cmd['offset'], cmd['total_size'])
        response['msg'] = 'cmd={} response={}'.format(cmd_name,
                                                      response['msg'])
        return response

    def write_reg(self, cmd_name, data, offset=0):
        """Writes a register defined by the memory map.

        Keyword arguments:
        offset -- the number of elements to offset in an array
        """
        cmd = self.cmd_list[cmd_name]
        response = None
        if cmd['is_bitfield'] is 'True':
            response = self.write_bits(cmd['offset'],
                                       cmd['bit_offset'],
                                       cmd['bits'], data)
        elif cmd['size'] != cmd['total_size']:
            offset = int(cmd['offset']) + (offset * int(cmd['size']))
            response = self.write_bytes(offset, data)
        else:
            response = self.write_bytes(cmd['offset'], data, int(cmd['size']))
        response['msg'] = 'cmd={} response={}'.format(cmd_name,
                                                      response['msg'])
        return response


def main():
    """Tests all functions supported functions."""
    from pprint import pprint
    logging.getLogger().setLevel(logging.DEBUG)

    mmif = LLMemMapIf("mem_map.csv", 'serial', '/dev/ttyUSB0')
    print('==================================================================')
    try:
        pprint(mmif.read_struct('sys'))
        pprint(mmif.read_reg('user_reg.64', 0, 10))
        pprint(mmif.write_reg('user_reg.64', [6, 6, 6], 2))
        pprint(mmif.write_reg('user_reg.64', [6, 6, 6], 2))
        pprint(mmif.read_reg('user_reg.64'))
    except KeyError:
        print("WARNING: Unrecognized mem_map")
    pprint(mmif.cmd_list)
    for cmd, val in mmif.cmd_list.items():
        # Supress unused variable warning
        val = val
        pprint(cmd)
        pprint(mmif.read_reg(cmd))


if __name__ == "__main__":
    main()
