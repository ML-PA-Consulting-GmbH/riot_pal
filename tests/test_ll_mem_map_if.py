"""Tests all functions supported functions."""
from pprint import pformat
from pathlib import Path
from serial import SerialException
from riot_pal import LLMemMapIf
import helpers


def try_connect(mem_map='mem_map.csv'):
    try:
        return LLMemMapIf(mem_map, 'serial', '/dev/ttyACM0', timeout=1)
    except SerialException:
        return LLMemMapIf(mem_map, 'serial', '/dev/ttyUSB0', timeout=1)


def test_philip_mem_map_basic(regtest):
    mmif = try_connect(str(Path(__file__).parents[1]) + '/mem_map/philip_mem_map.csv')
    regtest.write(pformat(mmif.cmd_list) + '\n')
    regtest.write(pformat(mmif.reset_mcu()) + '\n')
    regtest.write(pformat(mmif.execute_changes()) + '\n')
    mmif.close()


def test_philip_mem_map_read_reg(regtest):
    mmif = try_connect(str(Path(__file__).parents[1]) + '/mem_map/philip_mem_map.csv')
    regtest.write(pformat(mmif.reset_mcu()) + '\n')

    for cmd in sorted(mmif.cmd_list):
        # Supress unused variable warning
        cmd_vals = mmif.cmd_list[cmd]
        read_val = mmif.read_reg(cmd)
        if 'DEVICE_SPECIFIC' in cmd_vals['flag']:
            read_val.pop('data', None)
        elif cmd_vals['default'] is not '':
            assert helpers.weak_cmp(read_val['data'], cmd_vals['default'])
        regtest.write(pformat(read_val) + '\n')
    mmif.close()


def test_philip_mem_map_write_reg(regtest):
    mmif = try_connect(str(Path(__file__).parents[1]) + '/mem_map/philip_mem_map.csv')
    regtest.write(pformat(mmif.reset_mcu()) + '\n')

    for cmd in sorted(mmif.cmd_list):
        # Supress unused variable warning
        read_val = mmif.read_reg(cmd)
        write_ret = mmif.write_reg(cmd, helpers.try_add(read_val['data']))
        read_val = mmif.read_reg(cmd)
        regtest.write(pformat(write_ret) + '\n')
        regtest.write(pformat(read_val) + '\n')
    mmif.close()
