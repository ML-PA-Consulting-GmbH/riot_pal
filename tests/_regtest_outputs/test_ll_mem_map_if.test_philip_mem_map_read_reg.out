{'cmd': 'mcu_rst',
 'data': [' Initialized', ' Build Date: Jan 11 2019 17:29:28'],
 'msg': 'EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 393 2',
 'data': 0,
 'msg': 'cmd=adc[0].error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 392 1',
 'data': 0,
 'msg': 'cmd=adc[0].mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 400 8',
 'data': [0, 0, 0, 0, 0, 0, 0, 0],
 'msg': 'cmd=adc[0].res response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 395 1',
 'data': 0,
 'msg': 'cmd=adc[0].sample_rate response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 396 4',
 'data': 0,
 'msg': 'cmd=adc[0].value response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 409 2',
 'data': 0,
 'msg': 'cmd=adc[1].error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 408 1',
 'data': 0,
 'msg': 'cmd=adc[1].mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 416 8',
 'data': [0, 0, 0, 0, 0, 0, 0, 0],
 'msg': 'cmd=adc[1].res response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 411 1',
 'data': 0,
 'msg': 'cmd=adc[1].sample_rate response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 412 4',
 'data': 0,
 'msg': 'cmd=adc[1].value response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 322 2',
 'data': 0,
 'msg': 'cmd=i2c.clk_stretch_delay response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 342 4',
 'data': 0,
 'msg': 'cmd=i2c.f_ticks response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 320 1',
 'data': 0,
 'msg': 'cmd=i2c.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 320 1, read_bits 320 0 1',
 'data': 0,
 'msg': 'cmd=i2c.mode.addr_10_bit response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 320 1, read_bits 320 1 1',
 'data': 0,
 'msg': 'cmd=i2c.mode.general_call response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 320 1, read_bits 320 4 1',
 'data': 0,
 'msg': 'cmd=i2c.mode.nack_data response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 320 1, read_bits 320 2 1',
 'data': 0,
 'msg': 'cmd=i2c.mode.no_clk_stretch response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 320 1, read_bits 320 3 1',
 'data': 0,
 'msg': 'cmd=i2c.mode.reg_16_bit response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 328 1',
 'data': 0,
 'msg': 'cmd=i2c.r_count response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 330 4',
 'data': 0,
 'msg': 'cmd=i2c.r_ticks response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 346 6',
 'data': [0, 0, 0, 0, 0, 0],
 'msg': 'cmd=i2c.res response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 338 4',
 'data': 0,
 'msg': 'cmd=i2c.s_ticks response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 324 2',
 'data': 85,
 'msg': 'cmd=i2c.slave_addr_1 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 326 2',
 'data': 66,
 'msg': 'cmd=i2c.slave_addr_2 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 321 1',
 'data': 0,
 'msg': 'cmd=i2c.status response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 321 1, read_bits 321 1 1',
 'data': 0,
 'msg': 'cmd=i2c.status.AF response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 321 1, read_bits 321 2 1',
 'data': 0,
 'msg': 'cmd=i2c.status.BERR response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 321 1, read_bits 321 4 1',
 'data': 0,
 'msg': 'cmd=i2c.status.BUSY response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 321 1, read_bits 321 3 1',
 'data': 0,
 'msg': 'cmd=i2c.status.GENCALL response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 321 1, read_bits 321 0 1',
 'data': 0,
 'msg': 'cmd=i2c.status.OVR response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 321 1, read_bits 321 5 1',
 'data': 0,
 'msg': 'cmd=i2c.status.RSR response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 329 1',
 'data': 0,
 'msg': 'cmd=i2c.w_count response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 334 4',
 'data': 0,
 'msg': 'cmd=i2c.w_ticks response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 427 1',
 'data': 0,
 'msg': 'cmd=pwm.duty response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 425 2',
 'data': 0,
 'msg': 'cmd=pwm.error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 428 4',
 'data': 0,
 'msg': 'cmd=pwm.freq response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 424 1',
 'data': 0,
 'msg': 'cmd=pwm.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 432 8',
 'data': [0, 0, 0, 0, 0, 0, 0, 0],
 'msg': 'cmd=pwm.res response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 456 568',
 'data': 34,
 'msg': 'cmd=res response=ERANGE-Numerical result out of range [34]',
 'result': 'Error'}
{'cmd': 'rr 387 1',
 'data': 0,
 'msg': 'cmd=rtc.day_of_month response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 388 1',
 'data': 0,
 'msg': 'cmd=rtc.day_of_week response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 386 1',
 'data': 0,
 'msg': 'cmd=rtc.hour response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 385 1',
 'data': 0,
 'msg': 'cmd=rtc.minute response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 389 1',
 'data': 0,
 'msg': 'cmd=rtc.month response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 391 1',
 'data': [0],
 'msg': 'cmd=rtc.res response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 384 1',
 'data': 0,
 'msg': 'cmd=rtc.second response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 390 1',
 'data': 0,
 'msg': 'cmd=rtc.year response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 353 4',
 'data': 0,
 'msg': 'cmd=spi.error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 352 1',
 'data': 0,
 'msg': 'cmd=spi.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 357 11',
 'data': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 'msg': 'cmd=spi.res response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 287 1',
 'msg': 'cmd=sys.build_time.day_of_month response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 288 1',
 'msg': 'cmd=sys.build_time.day_of_week response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 286 1',
 'msg': 'cmd=sys.build_time.hour response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 285 1',
 'msg': 'cmd=sys.build_time.minute response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 289 1',
 'msg': 'cmd=sys.build_time.month response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 291 1',
 'data': [0],
 'msg': 'cmd=sys.build_time.res response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 284 1',
 'msg': 'cmd=sys.build_time.second response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 290 1',
 'msg': 'cmd=sys.build_time.year response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 297 1',
 'data': 0,
 'msg': 'cmd=sys.cr response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 297 1, read_bits 297 0 1',
 'data': 0,
 'msg': 'cmd=sys.cr.dut_rst response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 292 4',
 'data': 17061,
 'msg': 'cmd=sys.device_num response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 268 4',
 'data': 10010,
 'msg': 'cmd=sys.fw_rev response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 272 4',
 'data': 1,
 'msg': 'cmd=sys.if_rev response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 298 22',
 'data': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 'msg': 'cmd=sys.res response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 256 12',
 'msg': 'cmd=sys.sn response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 296 1',
 'data': 0,
 'msg': 'cmd=sys.status response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 296 1, read_bits 296 0 1',
 'data': 0,
 'msg': 'cmd=sys.status.update response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 276 8',
 'data': 0,
 'msg': 'cmd=sys.tick response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 443 1',
 'data': 0,
 'msg': 'cmd=tmr.duty response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 441 2',
 'data': 0,
 'msg': 'cmd=tmr.error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 444 4',
 'data': 0,
 'msg': 'cmd=tmr.freq response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 448 4',
 'data': 0,
 'msg': 'cmd=tmr.hi_us response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 452 4',
 'data': 0,
 'msg': 'cmd=tmr.lo_us response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 440 1',
 'data': 0,
 'msg': 'cmd=tmr.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 369 4',
 'data': 115200,
 'msg': 'cmd=uart.baud response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 377 1',
 'data': 0,
 'msg': 'cmd=uart.ctrl response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 377 1, read_bits 377 1 2',
 'data': 0,
 'msg': 'cmd=uart.ctrl.parity response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 377 1, read_bits 377 3 1',
 'data': 0,
 'msg': 'cmd=uart.ctrl.rts response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 377 1, read_bits 377 0 1',
 'data': 0,
 'msg': 'cmd=uart.ctrl.stop_bits response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 368 1',
 'data': 0,
 'msg': 'cmd=uart.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 379 5',
 'data': [0, 0, 0, 0, 0],
 'msg': 'cmd=uart.res response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 373 2',
 'data': 0,
 'msg': 'cmd=uart.rx_count response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 378 1',
 'data': 0,
 'msg': 'cmd=uart.status response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 378 1, read_bits 378 0 1',
 'data': 0,
 'msg': 'cmd=uart.status.cts response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 378 1, read_bits 378 2 1',
 'data': 0,
 'msg': 'cmd=uart.status.fe response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 378 1, read_bits 378 3 1',
 'data': 0,
 'msg': 'cmd=uart.status.nf response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 378 1, read_bits 378 4 1',
 'data': 0,
 'msg': 'cmd=uart.status.ore response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 378 1, read_bits 378 1 1',
 'data': 0,
 'msg': 'cmd=uart.status.pe response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 375 2',
 'data': 0,
 'msg': 'cmd=uart.tx_count response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 0 256',
 'data': [0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
          12,
          13,
          14,
          15,
          16,
          17,
          18,
          19,
          20,
          21,
          22,
          23,
          24,
          25,
          26,
          27,
          28,
          29,
          30,
          31,
          32,
          33,
          34,
          35,
          36,
          37,
          38,
          39,
          40,
          41,
          42,
          43,
          44,
          45,
          46,
          47,
          48,
          49,
          50,
          51,
          52,
          53,
          54,
          55,
          56,
          57,
          58,
          59,
          60,
          61,
          62,
          63,
          64,
          65,
          66,
          67,
          68,
          69,
          70,
          71,
          72,
          73,
          74,
          75,
          76,
          77,
          78,
          79,
          80,
          81,
          82,
          83,
          84,
          85,
          86,
          87,
          88,
          89,
          90,
          91,
          92,
          93,
          94,
          95,
          96,
          97,
          98,
          99,
          100,
          101,
          102,
          103,
          104,
          105,
          106,
          107,
          108,
          109,
          110,
          111,
          112,
          113,
          114,
          115,
          116,
          117,
          118,
          119,
          120,
          121,
          122,
          123,
          124,
          125,
          126,
          127,
          128,
          129,
          130,
          131,
          132,
          133,
          134,
          135,
          136,
          137,
          138,
          139,
          140,
          141,
          142,
          143,
          144,
          145,
          146,
          147,
          148,
          149,
          150,
          151,
          152,
          153,
          154,
          155,
          156,
          157,
          158,
          159,
          160,
          161,
          162,
          163,
          164,
          165,
          166,
          167,
          168,
          169,
          170,
          171,
          172,
          173,
          174,
          175,
          176,
          177,
          178,
          179,
          180,
          181,
          182,
          183,
          184,
          185,
          186,
          187,
          188,
          189,
          190,
          191,
          192,
          193,
          194,
          195,
          196,
          197,
          198,
          199,
          200,
          201,
          202,
          203,
          204,
          205,
          206,
          207,
          208,
          209,
          210,
          211,
          212,
          213,
          214,
          215,
          216,
          217,
          218,
          219,
          220,
          221,
          222,
          223,
          224,
          225,
          226,
          227,
          228,
          229,
          230,
          231,
          232,
          233,
          234,
          235,
          236,
          237,
          238,
          239,
          240,
          241,
          242,
          243,
          244,
          245,
          246,
          247,
          248,
          249,
          250,
          251,
          252,
          253,
          254,
          255],
 'msg': 'cmd=user_reg response=EOK-command success [0]',
 'result': 'Success'}
